### PR TITLE
[JENKINS-37332] - Prevent File descriptor leaks when reading manifests from JARs

### PR DESCRIPTION
**Pull request by oleg-nenashev**
Fixes a critical regression introduced in 2.16. File Descriptor leak happens, because JarUrlConnection does not bother to automatically close underlying resource input streams when we access them. Good blog post about it: http://www.genevski.com/2010/04/javaneturl-and-jarurlconnection-may.html

Cause:
- Initially the unstable code has been introduced by https://github.com/jenkinsci/jenkins/commit/19f9b63d74693d4de6f6906e9896399d18b4107f
- then this code has been transiently used by @stephenc in https://github.com/jenkinsci/jenkins/commit/2a450e5ac2ecd83a54f05c83729852073601816e , which caused locking of detached plugins and their further update

Changes:
- [x] Prevent file descriptor leak when reading manifests
- [x] Prevent potential file descriptor leak if we load bundles from another JAR (corner-case)
- [x] Improve diagnostics of file close operations
- [x] Unit tests

https://issues.jenkins-ci.org/browse/JENKINS-37332

@reviewbybees @daniel-beck  @jenkinsci/code-reviewers

### CODE REVIEW

**Comment by christ66** [comment]
🐝

File: core/src/test/java/hudson/PluginManagerTest.java [hunk 1]
```diff
@@ -68,4 +84,77 @@ public void parseInvalidRequestedPlugins() throws Exception {
...
         }
     }
+    
+    @Test
+    public void shouldProperlyParseManifestFromJar() throws IOException {
+        File jar = createHpiWithManifest();
+        final Manifest manifest = PluginManager.parsePluginManifest(jar.toURI().toURL());
+        
+        assertThat("manifest should have been read from the sample", manifest, notNullValue());
+        assertAttribute(manifest, "Created-By", "Apache Maven");
```
Comment (daniel-beck) [inline core/src/test/java/hudson/PluginManagerTest.java]: If you're testing that this is correctly parsing the manifest, maybe include a test for the line wrapped attributes?

Comment (oleg-nenashev) [inline core/src/test/java/hudson/PluginManagerTest.java]: done

**Comment by daniel-beck** [comment]
:+1:

**Comment by oleg-nenashev** [comment]
@reviewbybees done
@jenkinsci/code-reviewers I would like to get it integrated into the next weekly since it's a critical bug. Any concerns?
